File: apps/frontend/src/types/apiKey.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 TRUSCA contributors
/**
 * API Key wire types — chore C (`/integrations` UI).
 *
 * Mirrors apps/backend/schemas/api_key.py. Snake_case is preserved for the
 * wire shape so the JSON round-trip is verbatim; the React layer uses
 * these types directly without renaming because every other typed wrapper
 * in @/lib/projectsApi follows the same convention.
 */

export type APIKeyScope = "org" | "team" | "project";

/**
 * What a key may do, as distinct from the scope above, which says what it may
 * reach. A read-only key is refused every request that changes something, so
 * a pipeline that just reads results cannot start a scan.
 *
 * Keys issued before this existed are read-write and stay that way: narrowing
 * them on upgrade would have stopped whatever was already using them.
 */
export type APIKeyPermissionBreadth = "read_write" | "read_only";

export const API_KEY_PERMISSION_BREADTHS = [
  "read_only",
  "read_write",
] as const;

/** Response from POST /v1/api-keys — `raw_key` is shown ONCE. */
export interface APIKeyCreateOut {
  id: string;
  key_prefix: string;
  name: string;
  scope: APIKeyScope;
  permission_breadth: APIKeyPermissionBreadth;
  team_id: string | null;
  project_id: string | null;
  created_by_user_id: string | null;
  created_at: string;
  /** ISO timestamp when the key expires, or null when it never expires. */
  expires_at: string | null;
  /**
   * Plaintext bearer key (`tos_<prefix>_<secret>`). The backend returns this
   * exactly once at issuance; the SPA shows it in a "copy now" dialog and
   * never persists it. Subsequent reads return :class:`APIKeyListItem`.
   */
  raw_key: string;
}

/** Row shape returned by GET /v1/api-keys. NEVER includes the plaintext. */
export interface APIKeyListItem {
  id: string;
  key_prefix: string;
  name: string;
  scope: APIKeyScope;
  permission_breadth: APIKeyPermissionBreadth;
  team_id: string | null;
  project_id: string | null;
  created_by_user_id: string | null;
  /**
   * Email of the issuing user (L-17). Null when the issuer account was
   * deleted (created_by_user_id was SET NULL) — render an em-dash.
   */
  created_by_email: string | null;
  created_at: string;
  /** ISO timestamp when the key expires, or null when it never expires. */
  expires_at: string | null;
  last_used_at: string | null;
  revoked_at: string | null;
}

export interface APIKeyListPage {
  items: APIKeyListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface APIKeyCreatePayload {
  name: string;
  scope: APIKeyScope;
  /** Omitted means read-only, which is what the backend defaults new keys to. */
  permission_breadth?: APIKeyPermissionBreadth;
  /**
   * Issue the key to an automation identity rather than to the caller. The
   * key then lives as long as that identity does instead of stopping when the
   * person who issued it is deactivated. Omitted means a personal key.
   */
  service_account_id?: string | null;
  team_id?: string | null;
  project_id?: string | null;
  /**
   * Optional TTL in days (backend caps at 1–1825). Omitted / null → the key
   * never expires.
   */
  expires_in_days?: number | null;
}

export interface ListAPIKeysParams {
  scope?: APIKeyScope;
  team_id?: string;
  project_id?: string;
  include_revoked?: boolean;
  page?: number;
  page_size?: number;
}

File: apps/backend/schemas/api_key.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 TRUSCA contributors
"""
Pydantic schemas for API Key management — Phase 5 PR #16.

Public shapes:
  - APIKeyCreateIn        — request body for POST /v1/api-keys
  - APIKeyCreateOut       — response with the plaintext (returned ONCE)
  - APIKeyListItem        — list-row shape (no plaintext, no hash)
  - APIKeyListPage        — paginated list wrapper
  - APIKeyScope           — Literal type alias for the closed scope set

Design notes:
  - APIKeyCreateOut.raw_key is the plaintext bearer string
    (``tos_<prefix>_<secret>``). It is returned exactly once at issuance and
    intentionally NOT stored on the server side. The list endpoint NEVER echoes
    the plaintext — clients must capture it from the create response or rotate
    the key.
  - APIKeyListItem omits ``key_hash`` so a leaky serializer (e.g. a future bug
    that round-trips ORM rows directly) cannot accidentally surface the hash.
  - Literal types on ``scope`` give us crisp OpenAPI + Pydantic v2 validation;
    a bogus value fails fast with a 422 RFC 7807 envelope.
  - APIKeyListItem.last_used_at has interval resolution, not per-request
    resolution (concurrency-scaling-plan-2026-08-22.md A2). See its field
    description and ``core.config.api_key_last_used_at_update_interval_seconds``.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# Closed scope set. ``Literal`` is friendlier to OpenAPI than a Postgres ENUM
# round-trip — the API accepts only these three values.
APIKeyScope = Literal["org", "team", "project"]

# What a key may DO, as distinct from the scope above, which says what it may
# reach. Read-only keys are refused every unsafe HTTP method at the auth path.
APIKeyPermissionBreadth = Literal["read_write", "read_only"]
API_KEY_PERMISSION_BREADTHS: tuple[str, ...] = ("read_write", "read_only")


class APIKeyCreateIn(BaseModel):
    """Request body for creating a new API key.

    The CHECK constraint on the DB enforces scope coherence; we mirror it in
    Pydantic for fast client-side feedback. The router pre-validates that the
    actor can actually issue at the requested scope (super_admin → org,
    team_admin → team, team member → project).
    """

    name: str = Field(..., min_length=1, max_length=100)
    scope: APIKeyScope
    permission_breadth: APIKeyPermissionBreadth = Field(
        default="read_only",
        description=(
            "What the key may do: 'read_only' (the default) or 'read_write'. "
            "A read-only key is refused every request that changes something, "
            "so a pipeline that just reads results cannot start a scan. Keys "
            "issued before this existed are read-write and stay that way."
        ),
    )
    team_id: UUID | None = None
    project_id: UUID | None = None
    service_account_id: UUID | None = Field(
        default=None,
        description=(
            "Issue the key to an automation identity instead of to yourself. "
            "The key then lives as long as that identity does, rather than "
            "stopping when you are deactivated. Omit for a personal key, "
            "which keeps today's behaviour exactly."
        ),
    )
    expires_in_days: int | None = Field(
        default=None,
        ge=1,
        le=1825,  # 5 years
        description=(
            "Optional TTL in days. The key stops authenticating after this many "
            "days. Omit for a non-expiring key (CI keys should set one and "
            "rotate). Max 1825 (5 years)."
        ),
    )


class APIKeyNarrowIn(BaseModel):
    """Request body for PATCH /v1/api-keys/{id}.

    Only one value is accepted. Widening is not a validation failure to be
    argued with but a different operation: issue a new key. Spelling it as a
    literal keeps that in the OpenAPI contract rather than in a comment.
    """

    model_config = ConfigDict(extra="forbid")

    permission_breadth: Literal["read_only"] = Field(
        description=(
            "The only accepted value. Breadth narrows and never widens: a key "
            "that has been sitting in a CI log should not be handed more "
            "privilege than it was issued with."
        )
    )


class APIKeyCreateOut(BaseModel):
    """Response from POST /v1/api-keys.

    ``raw_key`` is the only place the plaintext is ever surfaced. The client
    is responsible for storing it (e.g. as a CI secret); a subsequent GET on
    the key returns the metadata-only :class:`APIKeyListItem` shape.
    """

    id: UUID
    key_prefix: str
    name: str
    scope: APIKeyScope
    permission_breadth: APIKeyPermissionBreadth = Field(
        description="What the key that was just issued may do."
    )
    team_id: UUID | None
    project_id: UUID | None
    created_by_user_id: UUID | None
    created_at: datetime
    expires_at: datetime | None = None
    raw_key: str = Field(
        ...,
        description=(
            "The plaintext bearer key (format: tos_<prefix>_<secret>). "
            "Returned exactly once at issuance; capture it client-side. "
            "Subsequent reads only return metadata."
        ),
    )


class APIKeyListItem(BaseModel):
    """List-row shape — never includes the plaintext or the hash."""

    id: UUID
    key_prefix: str
    name: str
    scope: APIKeyScope
    permission_breadth: APIKeyPermissionBreadth = Field(
        default="read_write",
        description=(
            "What this key may do. Rows issued before this existed read as "
            "'read_write', which is the breadth they have always had."
        ),
    )
    team_id: UUID | None
    project_id: UUID | None
    created_by_user_id: UUID | None
    created_by_email: str | None = Field(
        default=None,
        description=(
            "Email of the issuing user, so the management UI can show a "
            "human-readable creator column. None when the issuer account was "
            "deleted (created_by_user_id was SET NULL) or the user row is "
            "otherwise gone. PII note: this list is only reachable through the "
            "key-governance visibility boundary (issuer / team members / "
            "super_admin), so the email is not exposed beyond actors who can "
            "already manage the key."
        ),
    )
    created_at: datetime
    last_used_at: datetime | None = Field(
        default=None,
        description=(
            "When this key was last used to authenticate, rounded down to "
            "the nearest update interval (15 minutes by default, "
            "API_KEY_LAST_USED_AT_UPDATE_INTERVAL_SECONDS). This means "
            "'used at some point within that interval', not the exact "
            "instant of the most recent request. A key used twice inside "
            "one interval keeps the first commit's value. None means the "
            "key has never authenticated a request."
        ),
    )
    revoked_at: datetime | None
    expires_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class APIKeyListPage(BaseModel):
    """Paginated list of API keys."""

    items: list[APIKeyListItem]
    total: int
    page: int
    page_size: int


class APIKeyHashMigrationOut(BaseModel):
    """Response for ``GET /v1/admin/api-keys/hash-migration`` (A5).

    concurrency-scaling-plan-2026-08-22.md §3.3 A5: API-key hashing moved
    from bcrypt to a fast keyed HMAC-SHA256, expand-first. A key issued
    before that change keeps its bcrypt hash until it is reissued (there is
    no bulk-migration job), so this endpoint gives an operator the count
    needed to confirm "every active key has moved" before the follow-up
    change that drops bcrypt-hash reads from the auth path.
    """

    legacy_bcrypt_count: int = Field(
        description=(
            "Active (not revoked, not expired) API keys still hashed with "
            "the legacy bcrypt format. Zero means it is safe to schedule "
            "the contraction step (dropping bcrypt reads from the "
            "authentication path)."
        )
    )
    hmac_sha256_count: int = Field(
        description="Active API keys already hashed with the new HMAC-SHA256 format."
    )
    active_total: int = Field(description="legacy_bcrypt_count + hmac_sha256_count.")


__all__ = [
    "APIKeyCreateIn",
    "APIKeyCreateOut",
    "APIKeyHashMigrationOut",
    "APIKeyListItem",
    "APIKeyListPage",
    "APIKeyScope",
]
